src/NServiceBus.Core/Persistence/Learning/SagaPersister/SagaStorageFile.cs

```diff
@@ -0,0 +1,129 @@
+namespace NServiceBus
+{
+    using System;
+    using System.IO;
+    using System.Threading.Tasks;
+    using Janitor;
+
+    [SkipWeaving]
+    class SagaStorageFile : IDisposable
+    {
+        SagaStorageFile(FileStream fileStream, SagaManifest manifest)
+        {
+            this.fileStream = fileStream;
+            this.manifest = manifest;
+
+            lastModificationSeenAt = File.GetLastWriteTimeUtc(fileStream.Name);
+        }
+
+        public static bool TryOpen(Guid sagaId, SagaManifest manifest, out SagaStorageFile sagaStorageFile)
+        {
+            var filePath = manifest.GetFilePath(sagaId);
+
+            if (!File.Exists(filePath))
+            {
+                sagaStorageFile = null;
+                return false;
+            }
+
+            sagaStorageFile = new SagaStorageFile(new FileStream(filePath, FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite, DefaultBufferSize, FileOptions.Asynchronous), manifest);
+
+            return true;
+        }
+
+        public static SagaStorageFile Create(Guid sagaId, SagaManifest manifest)
+        {
+            var filePath = manifest.GetFilePath(sagaId);
+
+            return new SagaStorageFile(new FileStream(filePath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.ReadWrite, DefaultBufferSize, FileOptions.Asynchronous), manifest);
+        }
+
+        public void Dispose()
+        {
+            fileStream.Close();
+            fileStream.Dispose();
+
+            if (isCompleted)
+            {
+                File.Delete(fileStream.Name);
+            }
+
+            fileStream = null;
+        }
+
+        public Task Write(IContainSagaData sagaData)
+        {
+            fileStream.Position = 0;
+
+            return WriteAtomic((stream, state) =>
+            {
+                var serializer = state.Manifest.Serializer;
+                var data = state.SagaData;
+                serializer.WriteObject(stream, data);
+                return TaskEx.CompletedTask;
+            }, fileStream, lastModificationSeenAt, new WriteState(sagaData, manifest));
+        }
+
+        public async Task MarkAsCompleted()
+        {
+            await WriteAtomic((stream, state) => stream.WriteAsync(EmptyBytes, 0, 0), fileStream, lastModificationSeenAt)
+                .ConfigureAwait(false);
+
+            isCompleted = true;
+        }
+
+        public object Read()
+        {
+            return manifest.Serializer.ReadObject(fileStream);
+        }
+
+        static Task WriteAtomic(Func<FileStream, object, Task> action, FileStream stream, DateTime lastModificationSeenAt)
+        {
+            return WriteAtomic<object>(action, stream, lastModificationSeenAt);
+        }
+
+        static async Task WriteAtomic<TState>(Func<FileStream, TState, Task> action, FileStream stream, DateTime lastModificationSeenAt, TState state = default(TState))
```

Comment:
I wonder whether that should rather be named `WriteLocking` or something like that. It isn't an atomic and atomic writes are not intended to do concurrency checks as this method does.